server/controllers/user.controller.ts

```diff
@@ -1,30 +1,134 @@
             });
         }
 
-        const validUser = await this.verifyUser(req.body.password, user.password);
-        if (!validUser){
-            return res.status(400).json({
-                message: 'Invalid password!',
+        const validUser =  await this.registerUser(req.body.email, req.body.password, req.body.is_admin);
+        console.log(validUser.message);
+        if (! validUser.ok){
+            // verify the completed fields
+            return res.status(422).json({
+                message: 'Invalid input!'
             });
         }
 
-        const token = JWT.createAccessToken({ id: user.id }); //FIXME:  TypeError: Cannot read property 'createAccessToken' of undefined
+        else{
+            const email = req.body.email;
+            const encryptedPass = await bcrypt.hash(req.body.password, 10);
+            let isAdmin = req.body.is_admin;
+           
+            let profilePhoto = req.body.profile_photo;
+            if (profilePhoto == undefined) 
+                profilePhoto = "";
+
+            // insert the user into the table  
+            const SQL = `INSERT INTO public.user(email, password, is_admin, profile_photo) VALUES ('${email}', '${encryptedPass}', '${isAdmin}', '${profilePhoto}') RETURNING id`; 
+            
+            const pool = database.getPool();
+            const client = await pool.connect();
+
+            try {
+                const  insertedUser = await client.query(SQL);
+                var userId = insertedUser.rows[0].id;
+
+                console.log("User with id. " + userId + " added with success!");
+            }
+            catch(err){
+                console.error(err);
+            }
+            finally {
+                // Make sure to release the client before any error handling,
+                // just in case the error handling itself throws an error.
+                client.release()
+            }
+
+            const token = JWT.createAccessToken({ id: userId }); 
+            const refreshToken = JWT.createRefreshToken();
+
+            try {
+                JWT.storeRefreshToken(userId, refreshToken);    
+            } 
+            catch (err) {
+                console.log(err);
+
+                return res.status(500).json({
+                    message: 'An error occurred while signing up.',
+                });
+            }
+        
+            return res.status(200).json({
+                email: email,
+                token,
+                refreshToken,
+                id: userId
+            });
+
+            }  
+     } 
+
+     async loginUser(email: string, password: string){
+        let user = await this.getUserByEmail(email);
+    
+        if (!user)
+            return { message: "Invalid email!", user:user };
+         
+        const validUser =  await this.verifyUser(password, user.password);
+        if (!validUser)
+            return { message: "Invalid password!", user:user };
```

Comment:
From a security POV, I don't think it's recommended to tell exactly what the cause of failure was. Maybe _Invalid password or email_ could be used(this applies to line 106 as well)?

Reply:
I thought that specifying the exact cause will make the tests more accurate since I know exactly how the program works

Reply:
IMO we could leave it like that, but that would give significant information for an attacker - they would knew they've got the email right, all they have to do next is do try many attempts for the password.

As far as the testing part is concerned, I think you could test that regex directly. Maybe you could extract it into its own function, like `const isPasswordValid = pass => regex.test(pass)` and make more tests for it :thinking:.